py/query.py

```diff
@@ -0,0 +1,265 @@
+
+  def handle_request(self):
+    try:
+      request, client_address = self.get_request()
+    except socket.error:
+      return
+    if self.verify_request(request, client_address):
+      self.requests.put((request, client_address))
+
+#enable threaded server
+class ThreadedHTTPServer(ThreadPoolMixIn, HTTPServer):
+  pass
+
+#custom handler for getting routes
+class StoreHandler(BaseHTTPRequestHandler):
+
+  #boiler plate parsing
+  def parse_segments(self, post):
+    #split the query from the path
+    try:
+      split = urlparse.urlsplit(self.path)
+    except:
+      raise Exception('Try a url that looks like /action?query_string')
+    #path has the action in it
+    try:
+      if split.path.split('/')[-1] not in actions:
+        raise
+    except:
+      raise Exception('Try a valid action: ' + str([k for k in actions]))
+    #handle POST
+    if post:
+      body = self.rfile.read(int(self.headers['Content-Length'])).decode('utf-8')
+      return json.loads(body)
+    #handle GET
+    else:
+      params = urlparse.parse_qs(split.query)
+      if 'json' in params:      
+        return json.loads(params['json'][0])
+    raise Exception('No json provided')
+
+  #parse the request because we dont get this for free!
+  def handle_request(self, post):
+    #get the query data
+    segments = self.parse_segments(post)
+
+    try:   
+      # get the provider. 
+      list_of_ids = segments['segments'] if 'segments' in segments else None
+      start_date_time = segments.get('start_date_time', None)
+      end_date_time = segments.get('end_date_time', None)
+      cursor = self.server.sql_conn.cursor()
+
+      # send it to the cursor.
+
+      if None in (start_date_time, end_date_time):
+        cursor.execute("execute query_by_id (%s)",(list_of_ids,))
+      else:
+        s_date_time = calendar.timegm(time.strptime(start_date_time,"%Y-%m-%d"))
+        # add a day so that we include the end day in the range.
+        d = datetime.datetime.strptime(end_date_time,"%Y-%m-%d") + datetime.timedelta(days=1)
+        e_date_time = calendar.timegm(d.timetuple())
+
+        cursor.execute("execute query_by_range (%s, %s)",(s_date_time,e_date_time))
+      rows = cursor.fetchall()      
+      segments = {'segments':[]}
+
+      for row in rows:
+        d = dict()
+        d['segment_id'] = row[0]
+        d['prev_segment_id'] = row[1]
+        d['start_time'] = row[2]
+        d['start_time_dow'] = row[3]
+        d['start_time_hour'] = row[4]
+        d['end_time'] = row[5]
+        d['speed'] = row[6]
+        d['length'] = row[7]
+
+        segments['segments'].append(d)
+
+      results = json.dumps(segments)
+
+    except Exception as e:
+      # must commit if failure
+      #self.server.sql_conn.commit()             
+      return 400, str(e)
+
+    #hand it back
+    return 200, json.loads(results)
+
+  #send an answer
+  def answer(self, code, body):
+    response = json.dumps({'response': body })
```

Comment:
this should be `response = json.dumps(body, separators=(',', ':')) if type(body) == dict else json.dumps({'response': body}) `